In [4]:
# To run this code you need python 3 + the standard packages like numpy, scipy, matplotlib, pandas
# and a special package 'madx' that you can find at https://github.com/sterbini/cl2pd and you can install with
# pip install git+https://github.com/sterbini/madx.git
# The package is tested only on Mac. It should be working also on UNIX-like environment but it is not working on WIN.
from matplotlib import pyplot as plt
from madx import madx
from collections import OrderedDict

In [5]:
madExecutable = '/usr/local/bin/madx'
myMad = madx.MadX(madExecutable)

  ++++++++++++++++++++++++++++++++++++++++++++
  +     MAD-X 5.02.13  (64 bit, Darwin)      +
  + Support: mad@cern.ch, http://cern.ch/mad +
  + Release   date: 2016.12.20               +
  + Execution date: 2019.01.21 21:12:14      +
  ++++++++++++++++++++++++++++++++++++++++++++




# An introduction to MAD-X

In this first part we are going to get familiar with MAD-X syntax. This is an extension of what it is required from the tutorial.

For more information please refer to the [MAD-X online manual](http://cern.ch/madx/releases/last-rel/madxuguide.pdf).


In [6]:
myString='''
stop;
'''
myMad.input(myString);

stop;
  Number of warnings: 0
  ++++++++++++++++++++++++++++++++++++++++++++
  +          MAD-X finished normally         +
  ++++++++++++++++++++++++++++++++++++++++++++


With the 'stop;' instruction we exit from MAD-X, so, as done in the following cell we need to re-istantiate our MAD-X object with the 
**myMad = madx.MadX(madExecutable)** instruction.

---
It is a good practice to make header, please use '!' to comment the single line.

In [8]:
myMad = madx.MadX(madExecutable)
myString='''
!***************************************
! It is a good practice to make a header
!***************************************
title,'My test'; 
value, title;
'''
myMad.input(myString);

  ++++++++++++++++++++++++++++++++++++++++++++
  +     MAD-X 5.02.13  (64 bit, Darwin)      +
  + Support: mad@cern.ch, http://cern.ch/mad +
  + Release   date: 2016.12.20               +
  + Execution date: 2019.01.21 21:14:31      +
  ++++++++++++++++++++++++++++++++++++++++++++


!***************************************
! It is a good practice to make a header
!***************************************
title,'My test'; 
value, title;
title              =                  0 ;


---
Print the version of MAD-X

In [9]:
myString='''
value, VERSION;
'''
myMad.input(myString);

value, VERSION;
version            =              50213 ;


--- 
Use the **help** keyword (very rudimental help)

In [10]:
myString='''
help, title;
'''
myMad.input(myString);

help, title;
command: title  module: control
parameter: dummy   string: 0


---
This is an example to make a system call, just for showing you how to do it. The system call string (in this case 'pwd') depends on the Operating System.

In [13]:
myString='''
system,'date';
'''
myMad.input(myString);


system,'date';
Mon Jan 21 21:15:01 CET 2019


---
This is an example to get familiar with the use of the physical constants and the formatting of the output. Have a look on the difference.

In [14]:
myString='''
a=pi;
value a; 
set, format="22.20e";
value a; 
! reset to nominal values
option,reset=True;
set, format="10d", "18.10g", "-18s";
'''
myMad.input(myString);

a=pi;
value a; 
a                  =        3.141592654 ;
set, format="22.20e";
value a; 
a                  = 3.14159265358979311600e+00 ;
! reset to nominal values
option,reset=True;
set, format="10d", "18.10g", "-18s";


---
This is an example to get familiar with **if** and deferred expression. Please note the after the block delimited with **{...}** the **;** can be omitted. Pay attention to circular call!

In [15]:
myString='''
if (1==1){
option, echo=false, info=true;
a=pi;
b:=a;
c=a;
value a; 
value b;
value c;
a=CLIGHT*cos(a);
value a;
value b;
value c;}
! BEWARE of circular call!
!a:=a+1;
! When evaluating you will get a fatal error
! value a; 
option, echo=true, info=true;
'''
myMad.input(myString);

if (1==1){
option, echo=false, info=true;
a=pi;
b:=a;
c=a;
value a; 
value b;
value c;
a=CLIGHT*cos(a);
value a;
value b;
value c;}
++++++ info: a redefined
a                  =        3.141592654 ;
b                  =        3.141592654 ;
c                  =        3.141592654 ;
++++++ info: a redefined
a                  =         -299792458 ;
b                  =         -299792458 ;
c                  =        3.141592654 ;


---
This is an example to get familiar with **while** and **macros** loops.

In [16]:
myString='''
a(myvariable1,myvariable2): macro = {
value, myvariable1;
value, myvariable1*myvariable2;
}

N=1;
while (N<10){
exec, a(N,N);
N=N+1;
}
'''
myMad.input(myString);

a(myvariable1,myvariable2): macro = {
value, myvariable1;
value, myvariable1*myvariable2;
}
N=1;
while (N<10){
exec, a(N,N);
N=N+1;
}
n                  =                  1 ;
n*n                =                  1 ;
++++++ info: n redefined
n                  =                  2 ;
n*n                =                  4 ;
++++++ info: n redefined
n                  =                  3 ;
n*n                =                  9 ;
++++++ info: n redefined
n                  =                  4 ;
n*n                =                 16 ;
++++++ info: n redefined
n                  =                  5 ;
n*n                =                 25 ;
++++++ info: n redefined
n                  =                  6 ;
n*n                =                 36 ;
++++++ info: n redefined
n                  =                  7 ;
n*n                =                 49 ;
++++++ info: n redefined
n                  =                  8 ;
n*n                =                 64 ;
++++++ info: n rede

---
### List of functions
In MAD-X the following functions are available

- SQRT(x) square root,
- LOG(x) natural logarithm,
- LOG10(x) logarithm base 10,
- EXP(x) exponential,
- SIN(x) trigonometric sine,
- COS(x) trigonometric cosine,
- TAN(x) trigonometric tangent,
- ASIN(x) arc sine,
- ACOS(x) arc cosine,
- ATAN(x) arc tangent,
- SINH(x) hyperbolic sine,
- COSH(x) hyperbolic cosine,
- TANH(x) hyperbolic tangent,
- SINC(x) cardinal sine function,
- ABS(x) absolute value,
- ERF(x) Gauss error,
- ERFC(x) complementary error,
- FLOOR(x) floor, largest previous integer,
- CEIL(x) ceiling, smallest next integer,
- ROUND(x) round, closest integer,
- FRAC(x) fractional part of number,
- RANF() random number, uniformly distributed in [0,1],
- GAUSS() random number, gaussian distribution with unit standard deviation,
- TGAUSS(x) random number, gaussian distribution with unit standard deviation, truncated at x standard deviations;

In [17]:
myString='''
value, 1+1.2;
value, 1-1.2;
value, 1*1.2;
value, 1/1.2;
value, 2^0.5;
value, 1/0;
value, 2^-0.5;
value, sqrt(3);
value, log(10);
value, log10(10^2);
value, exp(0);
value, sin(pi/2);
value, cos(sqrt(2)/2);
value, tan(pi/2);
value, asin(0);
value, acos(0);
value, atan(0);
value, sinh(pi/2);
value, cosh(sqrt(2)/2);
value, tanh(pi/2);
value, sinc(0);
value, abs(-1.2);
value, erf(1);
value, 1-erfc(1);
value, ceil(2.3);
value, round(2.3);
value, frac(2.3);
value, ranf();
value, gauss();
value, tgauss(.1);
'''
myMad.input(myString)

value, 1+1.2;
1+1.2              =                2.2 ;
value, 1-1.2;
1-1.2              =               -0.2 ;
value, 1*1.2;
1*1.2              =                1.2 ;
value, 1/1.2;
1/1.2              =       0.8333333333 ;
value, 2^0.5;
2^0.5              =        1.414213562 ;
value, 1/0;
++++++ warning: division by zero, result set to zero, expr: 1/0
1/0                =                  0 ;
value, 2^-0.5;
++++++ warning: invalid expression: 2^-0.5
value, sqrt(3);
sqrt(3)            =        1.732050808 ;
value, log(10);
log(10)            =        2.302585093 ;
value, log10(10^2);
log10(10^2)        =                  2 ;
value, exp(0);
exp(0)             =                  1 ;
value, sin(pi/2);
sin(pi/2)          =                  1 ;
value, cos(sqrt(2)/2);
cos(sqrt(2)/2)     =       0.7602445971 ;
value, tan(pi/2);
tan(pi/2)          =    1.633123935e+16 ;
value, asin(0);
asin(0)            =                  0 ;
value, acos(0);
acos(0)            =        1.570796327 ;
value, a

---
### List of physical constant

| MAD-X name  | symbol  |  value |unit|
|:-:|:-:|:-:|:-:|
|PI| π |4 * atan(1)| 1|
|TWOPI|2π| 2 * PI| 1|
|DEGRAD| 180/π |180 / PI| deg/rad|
|RADDEG| π/180 |PI / 180 |rad/deg|
|E| e |exp(1) |1|
|EMASS| me |0.510998928e−3| GeV|
|PMASS| mp |0.938272046| GeV|
|NMASS| u |0.931494061| GeV|
|MUMASS| mµ| 0.1056583715 |GeV|
|CLIGHT| c| 299792458| m/s|
|QELECT| e| 1.602176565e−19| A.s|
|HBAR| ¯h| 6.58211928e−25| MeV.s|
|ERAD| re| 2.8179403267e−15| m|
|PRAD| re(me/mp)| ERAD*EMASS/PMASS| m|


In [18]:
myString='''
value, pi;
value, twopi;
value, raddeg;
value, degrad;
value, e;
value, emass;
value, pmass;
value, nmass;
value, mumass;
value, clight;
value, qelect;
value, hbar;
value, erad;
value, prad;
'''
myMad.input(myString)

value, pi;
pi                 =        3.141592654 ;
value, twopi;
twopi              =        6.283185307 ;
value, raddeg;
raddeg             =      0.01745329252 ;
value, degrad;
degrad             =        57.29577951 ;
value, e;
e                  =        2.718281828 ;
value, emass;
emass              =     0.000510998928 ;
value, pmass;
pmass              =        0.938272046 ;
value, nmass;
nmass              =        0.931494061 ;
value, mumass;
mumass             =       0.1056583715 ;
value, clight;
clight             =          299792458 ;
value, qelect;
qelect             =    1.602176565e-19 ;
value, hbar;
hbar               =     6.58211928e-25 ;
value, erad;
erad               =    2.817940327e-15 ;
value, prad;
prad               =     1.53469827e-18 ;
